Add support for cvfmsbundles file

RohitTopi opened the pull request:
Support for .cvmfsbundles files

To improve cold cache loading, a new content type, bundle, is introduced
([PR 2741](https://github.com/cvmfs/cvmfs/pull/2741)). The bundle has a binary format of an object pack

A bundle specification file, named `.cvmfsbundles`, allows the creation of such
bundles on the server-side. The `.cvfmsbundles` file is maintained by the
repository owners. During publication, the bundles are created and
stored according to the contents of this file.

This PR will
 * allow recognition of the bundle specification files on the server-side.
 * enforce the requirement that such files must be regular and not a hard link.
     Additionally, ensure that the file is in the root directory of a repository.
 * reserve the filename, `.cvmfsbundles`, for the bundle specification files. 

Corresponding JIRA issue: [CVM-1886](https://sft.its.cern.ch/jira/browse/CVM-1886)

RohitTopi committed:
Update sync_mediator.cc

added a check

RohitTopi committed:
Update sync_item.h

added function isBundleSpec()

RohitTopi committed:
Update sync_mediator.cc

renamed function isBundlesFile() to isBundleSpec()

RohitTopi committed:
Update sync_mediator.cc

fixed finding .cvmfsbundles files

RohitTopi committed:
Update sync_item.h

added wasBundleSpec()

RohitTopi committed:
Update sync_mediator.cc

added an if check for cvmfsbundles file in SyncMediator::Remove function

RohitTopi committed:
changed add function

RohitTopi committed:
Merge branch 'cvmfs:devel' into cvmfsbundlesinit

RohitTopi committed:
read and display cvmfsbundles file as JSON

RohitTopi committed:
added empty test

RohitTopi committed:
Merge branch 'cvmfs:devel' into cvmfsbundlesinit

RohitTopi committed:
Merge branch 'cvmfsbundlesinit' of https://github.com/RohitTopi/cvmfs into cvmfsbundlesinit

mcb30 commented:
Is this a work-in-progress PR that was submitted accidentally?  The code appears not to actually do anything.

jblomer commented on cvmfs/sync_mediator.cc:
```diff
@@ -198,6 +223,13 @@ void SyncMediator::Remove(SharedPtr<SyncItem> entry) {
     return;
   }
+  
+  if(entry->wasBundleSpec()) {
```
```suggestion
  if (entry->wasBundleSpec()) {
```

jblomer commented on cvmfs/sync_item.h:
```diff
@@ -93,6 +93,12 @@ class SyncItem {
            WasType(kItemFifo) ||
            WasType(kItemSocket);
   }
+  inline bool isBundleSpec() const {
```
```suggestion
  inline bool IsBundleSpec() const {
```

jblomer commented on cvmfs/sync_item.h:
```diff
@@ -93,6 +93,12 @@ class SyncItem {
+  inline bool isBundleSpec() const {
+    return filename_ == ".cvmfsbundles";
+  }
+  inline bool wasBundleSpec() const {
```
```suggestion
  inline bool WasBundleSpec() const {
```

RohitTopi commented:
Hello @mcb30
Yes, this is a work in progress PR. I've changed the current PR status to draft

jblomer commented on cvmfs/sync_mediator.cc:
```diff
@@ -93,6 +94,30 @@ void SyncMediator::Add(SharedPtr<SyncItem> entry) {
   }
+  
+  // .cvmfsbundles file type
+  if(entry->isBundleSpec()){
```
```suggestion
  if (entry->isBundleSpec()) {
```

jblomer commented on cvmfs/sync_mediator.cc:
```diff
@@ -93,6 +94,30 @@ void SyncMediator::Add(SharedPtr<SyncItem> entry) {
+    std::string json_string;
+
+    int fd = open (entry->GetUnionPath().c_str(), O_RDONLY);
+    if(fd >= 0) {
```
Let's use `PANIC` here, like
```
if (fd < 0) {
  PANIC(...)
}
```

jblomer commented on cvmfs/sync_mediator.cc:
```diff
@@ -93,6 +94,30 @@ void SyncMediator::Add(SharedPtr<SyncItem> entry) {
+
+    std::string json_string;
+
+    int fd = open (entry->GetUnionPath().c_str(), O_RDONLY);
```
```suggestion
    int fd = open(entry->GetUnionPath().c_str(), O_RDONLY);
```

jblomer commented on cvmfs/sync_mediator.cc:
```diff
@@ -93,6 +94,30 @@ void SyncMediator::Add(SharedPtr<SyncItem> entry) {
+
+    int fd = open (entry->GetUnionPath().c_str(), O_RDONLY);
+    if(fd >= 0) {
+      if(SafeReadToString(fd, &json_string)) {
```
Same here, let's PANIC if reading fails.
```suggestion
      if (SafeReadToString(fd, &json_string)) {
```

jblomer commented on cvmfs/sync_mediator.cc:
```diff
@@ -121,7 +146,7 @@ void SyncMediator::Add(SharedPtr<SyncItem> entry) {
     return;
   }
-
+  
```
Remove spaces

jblomer reviewed (changes requested):
That's a good start! Please add the other PRs (as drafts) as well so that we get a complete picture.  I added some style fixes.

RohitTopi committed:
Update cvmfs/sync_mediator.cc

Co-authored-by: Jakob Blomer <jblomer@cern.ch>

RohitTopi committed:
Update cvmfs/sync_item.h

Co-authored-by: Jakob Blomer <jblomer@cern.ch>

RohitTopi committed:
Update cvmfs/sync_mediator.cc

Co-authored-by: Jakob Blomer <jblomer@cern.ch>

RohitTopi committed:
Update cvmfs/sync_item.h

Co-authored-by: Jakob Blomer <jblomer@cern.ch>

RohitTopi committed:
Update cvmfs/sync_mediator.cc

Co-authored-by: Jakob Blomer <jblomer@cern.ch>

RohitTopi committed:
Update cvmfs/sync_mediator.cc

Co-authored-by: Jakob Blomer <jblomer@cern.ch>

RohitTopi committed:
Update sync_mediator.cc

RohitTopi committed:
Update sync_mediator.cc

RohitTopi committed:
Update sync_mediator.cc

RohitTopi commented:
Thank you, I've added the other draft PR's

RohitTopi committed:
Remove unnecessary printf calls

RohitTopi committed:
Remove unnecessary printf call

RohitTopi committed:
Style fixes

jblomer commented on cvmfs/sync_mediator.cc:
```diff
@@ -94,6 +95,28 @@ void SyncMediator::Add(SharedPtr<SyncItem> entry) {
+    int fd = open(entry->GetUnionPath().c_str(), O_RDONLY);
+    if (fd < 0) {
+      PANIC(kLogStderr, "Could not open file: %s",
+          entry->GetUnionPath().c_str());
```
```suggestion
            entry->GetUnionPath().c_str());
```

jblomer commented on cvmfs/sync_mediator.cc:
```diff
@@ -94,6 +95,28 @@ void SyncMediator::Add(SharedPtr<SyncItem> entry) {
+    if (fd < 0) {
+      PANIC(kLogStderr, "Could not open file: %s",
+          entry->GetUnionPath().c_str());
+    } else {
```
We don't need the `else` branch because we `PANIC` out otherwise

jblomer commented on cvmfs/sync_mediator.cc:
```diff
@@ -94,6 +95,28 @@ void SyncMediator::Add(SharedPtr<SyncItem> entry) {
+      PANIC(kLogStderr, "Could not open file: %s",
+          entry->GetUnionPath().c_str());
+    } else {
+      if (SafeReadToString(fd, &json_string)) {
```
Let's reverse: `if (!SafeReadToString(...` then panic, in which case we again can drop the else branch.

jblomer commented on cvmfs/sync_mediator.cc:
```diff
@@ -94,6 +95,28 @@ void SyncMediator::Add(SharedPtr<SyncItem> entry) {
+          entry->GetUnionPath().c_str());
+    } else {
+      if (SafeReadToString(fd, &json_string)) {
+        JsonDocument* json = JsonDocument::Create(json_string);
```
```suggestion
        JsonDocument *json = JsonDocument::Create(json_string);
```

jblomer commented on cvmfs/sync_mediator.cc:
```diff
@@ -94,6 +95,28 @@ void SyncMediator::Add(SharedPtr<SyncItem> entry) {
 
+  // .cvmfsbundles file type
+  if (entry->IsBundleSpec()) {
+    std::string json_string;
```
We should ensure that the file is a regular file (not a "special file") and not a hardlink.  Also that the file is in the root directory.  Otherwise we panic.

jblomer commented on test/src/684-cvmfsbundles/main:
```diff
@@ -0,0 +1,15 @@
+  echo "*** Create a new repository"
+  create_empty_repo $CVMFS_TEST_REPO $CVMFS_TEST_USER || return 1
+
+
```
We should already add code here to create .cvmfsbundles file, edit it, and remove it.  Also, we should already exercise failing malformed bundle files: a .cvmfsbundles directory and a .cvmfsbundles file in a subdirectory.

jblomer reviewed (changes requested):
It's getting there!  A few smaller comments.

RohitTopi committed:
Style fixes

RohitTopi committed:
Remove redundant else branches

RohitTopi committed:
Ensure that .cvmfsbundles file is of proper type

.cvmfsbundles file must be a regular file and not a hardlink

RohitTopi committed:
Ensure that .cvmfsbundles file is in root directory

RohitTopi committed:
Add test for .cvmfsbundles file

Test for creating, editing and removing a .cvmfsbundles file

RohitTopi committed:
Remove duplicate lines of code

RohitTopi committed:
Add test for .cvmfsbundles file

Test for failure in case a .cvmfsbundles file exists in a subdirectory
 of a repository

jblomer commented on cvmfs/sync_mediator.cc:
```diff
@@ -94,6 +95,33 @@ void SyncMediator::Add(SharedPtr<SyncItem> entry) {
   }
 
+  // .cvmfsbundles file type
+  if (entry->IsBundleSpec() && entry->IsRegularFile() &&
```
We should be more strict here: with this code, if we have a "weird" .cvmfsbundle file we'll ignore the special name.  However, we should check for `IsBundleSpec()` and if it is weird (no regular file, hardlink, etc.) then we abort.

jblomer commented on cvmfs/sync_mediator.cc:
```diff
@@ -94,6 +95,33 @@ void SyncMediator::Add(SharedPtr<SyncItem> entry) {
+  if (entry->IsBundleSpec() && entry->IsRegularFile() &&
+      !entry->HasHardlinks()) {
+    std::string parent_path = GetParentPath(entry->GetUnionPath());
+    if (!IsMountPoint(parent_path)) {
```
`IsMountPoint` probably works but is not the correct check.  We should rather check `parent_path != SyncUnion::union_path()`

jblomer reviewed (changes requested):
Some comments added.  Let's also add a big warning if we encounter a bundles file that tells the user that bundles is currently an experimental feature.

RohitTopi committed:
Improve checking of .cvmfsbundles file type

If a .cvmfsbundles file exists then it must be a regular file and not a
hardlink

RohitTopi committed:
Replace IsMountPoint() with SyncUnion::union_path()

RohitTopi committed:
Add warning about the experimental bundle feature

RohitTopi committed:
Disallow directories named .cvmfsbundles

Reserve .cvmfsbundles filename for bundles specification files

RohitTopi committed:
Add test for malformed .cvmfsbundles file

Test for failure in case a .cvmfsbundles directory exists in a
repository

RohitTopi committed:
Remove redundant comment

jblomer commented on cvmfs/sync_mediator.cc:
```diff
@@ -972,6 +1014,12 @@ void SyncMediator::AddUnmaterializedDirectory(SharedPtr<SyncItem> entry) {
 }
 
 void SyncMediator::AddDirectory(SharedPtr<SyncItem> entry) {
+  if (entry->filename() == ".cvmfsbundles") {
```
We should avoid using the string literal. Can we check instead for `IsBundleSpec()`?

jblomer reviewed (changes requested):
Very good! One more small comment and then this is good to go.

jblomer commented on cvmfs/sync_mediator.cc:
```diff
@@ -94,6 +95,41 @@ void SyncMediator::Add(SharedPtr<SyncItem> entry) {
+      PANIC(kLogStderr, "Could not read contents of file: %s",
+            entry->GetUnionPath().c_str());
+    }
+    JsonDocument *json = JsonDocument::Create(json_string);
```
```suggestion
    UniquePtr<JsonDocument> json(JsonDocument::Create(json_string));
```

RohitTopi committed:
Replace pointer with UniquePtr

RohitTopi committed:
Avoid using string literal

Use IsBundleSpec() function to check if the directory has the name
.cvmfsbundles

RohitTopi commented:
Thank you for the review. I have resolved all the comments and updated the description.

@jblomer
@ebocchi

jblomer commented:
@cernvm-bot cloudtest
